src/core/ext/filters/client_channel/client_channel.cc

```diff
@@ -545,23 +553,56 @@ static void on_resolver_result_changed_locked(void* arg, grpc_error* error) {
         gpr_log(GPR_INFO, "chand=%p: updating existing LB policy \"%s\" (%p)",
                 chand, lb_policy_name.get(), chand->lb_policy.get());
       }
-      chand->lb_policy->UpdateLocked(*chand->resolver_result);
+      // case (b) or (c)
+      trace_this_address_resolution =
+          chand->lb_policy->UpdateLocked(*chand->resolver_result);
       // No need to set the channel's connectivity state; the existing
       // watch on the LB policy will take care of that.
       set_connectivity_state = false;
     } else {
+      trace_this_address_resolution = true;  // case (d)
       // Instantiate new LB policy.
       create_new_lb_policy_locked(chand, lb_policy_name.get(),
                                   &connectivity_state, &connectivity_error);
+      // we also log the name of the new LB policy in addition to logging this
+      // resolution event.
+      if (chand->channelz_channel != nullptr) {
+        char* str;
+        gpr_asprintf(&str, "Switched LB policy to %s", lb_policy_name.get());
+        chand->channelz_channel->AddTraceEvent(
+            grpc_core::channelz::ChannelTrace::Severity::Info,
+            grpc_slice_from_copied_string(str));
+        gpr_free(str);
+      }
     }
     // Find service config.
     grpc_core::UniquePtr<char> service_config_json =
         get_service_config_from_resolver_result_locked(chand);
+    if ((service_config_json == nullptr &&
+         chand->info_service_config_json != nullptr) ||
+        (service_config_json != nullptr &&
+         chand->info_service_config_json == nullptr) ||
+        (service_config_json != nullptr &&
+         chand->info_service_config_json != nullptr &&
+         gpr_stricmp(service_config_json.get(),
```

Comment:
I think this comparison should be case-sensitive, since there could be fields whose values are case-sensitive.  And in most situations, the case won't change randomly anyway, so there probably won't be false positives if we do a case-sensitive check.

Reply:
Doe